src/pseudo/PspUpf.jl

```diff
@@ -190,6 +214,42 @@ function eval_psp_local_fourier(psp::PspUpf, q::T)::T where {T <: Real}
     4T(π) * (s - psp.Zion / q) / q
 end
 
+function eval_psp_density_valence_real(psp::PspUpf, r::T) where {T <: Real}
+    psp.r2_4π_ρion_interp(r) / (r^2 * 4T(π))
+end
+
+"""
+    eval_psp_density_valence_fourier(psp::PspUpf, q<:Real)
+
+For UPFs, the integral is transformed into the following sum:
+Σ{i} j_0(q r[i]) r^2 4π ρval[i] dr[i]
+"""
+function eval_psp_density_valence_fourier(psp::PspUpf, q::T) where {T <: Real}
+    s = zero(T)
+    @inbounds for ir = eachindex(psp.r2_4π_ρion_dr)
+        s += sphericalbesselj_fast(0, q * psp.rgrid[ir]) * psp.r2_4π_ρion_dr[ir]
+    end
+    s
+end
+
+function eval_psp_density_core_real(psp::PspUpf, r::T) where {T <: Real}
+    psp.ρcore_interp(r)
+end
+
+"""
+    eval_psp_density_core_fourier(psp::PspUpf, q<:Real)
+
+For UPFs, the integral is transformed into the following sum:
+4π Σ{i} j_0(q r[i]) r^2 ρcore[i] dr[i]
+"""
```

Comment:
should not be a docstring

Reply:
(abstract methods should be documented once in their abstract form, implementation-specific details should just be comments)